src/microbe_stage/MicrobeSystem.cs

```diff
@@ -54,4 +57,63 @@ public void Process(float delta)
             microbe.ProcessSync(delta);
         }
     }
+
+    /// <summary>
+    ///   Tries to find specified Species as close to the point as possible.
+    /// </summary>
+    /// <param name="position">Position to search around</param>
+    /// <param name="species">What species to search for</param>
+    /// <param name="searchRadius">How wide to search around the point</param>
+    /// <returns>The nearest found point for the species or null</returns>
+    public (Microbe Microbe, Vector3 Position)? FindSpeciesNearPoint(Vector3 position, Species species,
+        float searchRadius = 200)
+    {
+        if (searchRadius < 1)
+            throw new ArgumentException("searchRadius must be >= 1");
+
+        (Microbe Microbe, Vector3 Position)? closestMicrobe = null;
+        float nearestDistanceSquared = float.MaxValue;
+        var searchRadiusSquared = searchRadius * searchRadius;
+        microbes ??= worldRoot.GetTree().GetNodesInGroup(Constants.RUNNABLE_MICROBE_GROUP).Cast<Microbe>()
+            .ToArray();
+
+        foreach (var microbe in microbes)
+        {
+            if (microbe.Species != species)
+                continue;
+
+            Vector3 microbeGlobalPosition;
+
+            // Use colony parent position to avoid calling GlobalTranslation
+            if (microbe.ColonyParent != null)
+            {
+                microbeGlobalPosition = microbe.ColonyParent.Translation;
+            }
+            else
+            {
+                microbeGlobalPosition = microbe.Translation;
+            }
+
+            // Skip candidates for performance
```

Comment:
Does this *actually* perform better? This is multiple branching operations whereas the distance check should be *just* about 20 float operations and a single comparison.

Reply:
Ignoring data instructions and if multiplication takes 2 ops, this roughly saves 7 ops per miss and adds 6 ops per hit. Since the vast majority of calls will be misses i think it is faster. However the difference might be negligible.

Reply:
Isn't any branching operation slower than even division? That's my point that optimizing a bit of math is always slower on a modern CPU as branching is so expensive operation (hence why branchless programming is such a popular performance topic).

Reply:
The branch prediction will likely prefetch the `continue` path since it is the most likely. Then we need to add the cost of flushing the pipline to the added cost on a hit (lets say 10 cycles). So the change in cycles when using the branching is\
`(1-hitrate) * -7  + hitrate * 16` which is negative for a hitrate smaller than 30%. Though some profiling is probably needed to get a real comparison.

Reply:
tldr I think the branch is pretty deterministic and thus will have only a small performance impact.

Reply:
Well I'd be really interested to show what a benchmark done using https://github.com/dotnet/BenchmarkDotNet (I've seen this used by serious C# performance related Stackoverflow answers use) would show in this case.

Reply:
So i built the [test bench](https://github.com/ottoblep/thrive-profiling/blob/master/Program.cs) (BenchmarkDotNet works really well) and got these results. If the search radius is large compared to the spawn radius and if there are many of the searched species then the NoBranch option is better otherwise the reverse is true. In the average in-game case i got from testing in the editor (FRAC=0.05, RADIUS=150) the Branch option is slightly better.
```
|                       Method | SPECIES_FRACTION | SEARCH_RADIUS  |     Mean |    Error |   StdDev |   Median |
|----------------------------- |----------------- |--------------- |---------:|---------:|---------:|---------:|
|   FindSpeciesNearPointBranch |             0.05 |            50  | 617.9 ns |  1.49 ns |  1.32 ns | 618.1 ns |
| FindSpeciesNearPointNoBranch |             0.05 |            50  | 663.6 ns | 19.81 ns | 54.24 ns | 645.8 ns |
|   FindSpeciesNearPointBranch |             0.05 |           150  | 638.7 ns | 12.24 ns | 14.09 ns | 634.6 ns |
| FindSpeciesNearPointNoBranch |             0.05 |           150  | 647.1 ns | 12.72 ns | 22.62 ns | 637.6 ns |
|   FindSpeciesNearPointBranch |             0.05 |           350  | 642.0 ns | 12.64 ns | 17.72 ns | 635.3 ns |
| FindSpeciesNearPointNoBranch |             0.05 |           350  | 647.6 ns | 12.91 ns | 21.57 ns | 645.9 ns |
|   FindSpeciesNearPointBranch |              0.2 |            50  | 629.8 ns | 10.44 ns | 12.42 ns | 624.5 ns |
| FindSpeciesNearPointNoBranch |              0.2 |            50  | 615.8 ns |  2.12 ns |  1.77 ns | 615.9 ns |
|   FindSpeciesNearPointBranch |              0.2 |           150  | 622.5 ns |  3.96 ns |  3.51 ns | 621.5 ns |
| FindSpeciesNearPointNoBranch |              0.2 |           150  | 618.0 ns |  4.67 ns |  4.37 ns | 616.6 ns |
|   FindSpeciesNearPointBranch |              0.2 |           350  | 620.2 ns |  1.60 ns |  1.33 ns | 620.2 ns |
| FindSpeciesNearPointNoBranch |              0.2 |           350  | 615.3 ns |  2.54 ns |  1.98 ns | 615.4 ns |
|   FindSpeciesNearPointBranch |              0.5 |            50  | 623.7 ns |  4.17 ns |  3.70 ns | 623.1 ns |
| FindSpeciesNearPointNoBranch |              0.5 |            50  | 614.4 ns |  0.80 ns |  0.67 ns | 614.2 ns |
|   FindSpeciesNearPointBranch |              0.5 |           150  | 621.9 ns |  3.68 ns |  3.26 ns | 621.0 ns |
| FindSpeciesNearPointNoBranch |              0.5 |           150  | 616.1 ns |  2.44 ns |  2.28 ns | 615.1 ns |
|   FindSpeciesNearPointBranch |              0.5 |           350  | 623.1 ns |  4.74 ns |  4.21 ns | 621.9 ns |
| FindSpeciesNearPointNoBranch |              0.5 |           350  | 615.7 ns |  1.31 ns |  1.09 ns | 615.1 ns |
```

Reply:
Thanks for getting the actual numbers. 

I suspect that most of the time the search radius will be quite high, and as there are planned to be roughly 10 species per patch (this is a tweak value in auto-evo), the conclusion is that the branch version is slightly faster usually but sometimes a bit slower. So I suppose this is actually a very slight optimization which means this optimization can be kept.